src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java

```diff
@@ -116,33 +124,86 @@
         final DetailAST firstSwitchMemberAst = findFirstSwitchMember(ast);
 
-        if (!containsDefaultSwitch(firstSwitchMemberAst) && !isSwitchExpression(ast)) {
+        if (!containsDefaultLabel(firstSwitchMemberAst)
+                && !containsPatterCaseLabelElement(firstSwitchMemberAst)
+                && !containsNullReferenceCaseLabelElement(firstSwitchMemberAst)
+                && !containsDefaultCaseLabelElement(firstSwitchMemberAst)
+                && !isSwitchExpression(ast)) {
             log(ast, MSG_KEY);
         }
     }
 
     /**
-     * Checks if the case group or its sibling contain the 'default' switch.
+     * Checks if a switch block contains a case label with a pattern variable definition.
+     * In this situation, the compiler enforces the given switch block to cover
+     * all possible inputs, and we do not need a default label.
      *
      * @param caseGroupAst first case group to check.
-     * @return true if 'default' switch found.
+     * @return true if switch block contains a pattern case label element
      */
-    private static boolean containsDefaultSwitch(DetailAST caseGroupAst) {
+    private static boolean containsPatterCaseLabelElement(DetailAST caseGroupAst) {
+        final Predicate<DetailAST> astPredicate = ast -> {
+            return ast.getFirstChild() != null
+                    && ast.getFirstChild().findFirstToken(TokenTypes.PATTERN_VARIABLE_DEF) != null;
+        };
+        return findAstByPredicate(caseGroupAst, astPredicate) != null;
+    }
+
+    /**
+     * Checks if a switch block has case label with null reference. In this situation,
+     * the compiler enforces the given switch block to cover all possible inputs,
+     * and we do not need a default label.
+     *
+     * @param caseGroupAst first case group to check.
+     * @return true if switch block contains a null reference
+     */
+    private static boolean containsNullReferenceCaseLabelElement(DetailAST caseGroupAst) {
         DetailAST nextAst = caseGroupAst;
         boolean found = false;
 
-        while (nextAst != null) {
-            if (nextAst.findFirstToken(TokenTypes.LITERAL_DEFAULT) != null) {
+        while (nextAst != null && nextAst.getFirstChild() != null) {
+            final DetailAST grandChild = nextAst.getFirstChild().getFirstChild();
+            final DetailAST nullReference = findAstByPredicate(grandChild,
+                    ast -> ast.findFirstToken(TokenTypes.LITERAL_NULL) != null
+            );
+
+            if (nullReference != null) {
                 found = true;
                 break;
             }
-
             nextAst = nextAst.getNextSibling();
         }
 
         return found;
     }
 
+    /**
+     * Checks if the case group or its sibling contain the 'default' switch.
+     *
+     * @param caseGroupAst first case group to check.
+     * @return true if 'default' switch found.
+     */
+    private static boolean containsDefaultLabel(DetailAST caseGroupAst) {
```

Comment:
Changed the name of this method to be consistent with language from the JLS: https://docs.oracle.com/javase/specs/jls/se17/preview/specs/patterns-switch-jls.html